### PR TITLE
Animate activestate on frameless dialog boxes.

### LINKED ISSUE #912: Create "Active" state for Export and Print buttons
_Child issue w/in #910_

## Punchlist This Issue
_[Zeplin, here](https://zpl.io/VKgpODA)_ 
(note: this mockup uses screenshots w/ graphics added where things changed, so if you're confused why some text looks odd—well, that's why!)
- [ ] Add active-state to all primary buttons across flow(s)
- [ ] When re-submitting a page with an error on it, either:
   - a: Clear-out the error to return the screen to its state, or
   - b: If "a" would cause the height of the screen to shorten (because the error made the screen taller) just remove the error and show whitespace. We aim for this to be an epilepsy-free experience. :D

### LINKED ISSUE #927: Export/Print "Speedbump" screens experience polish
_Child issue w/in #910_

## This Issue's Punchlist
_[Zeplin, here](https://zpl.io/2vp1xmr)_
(note: this mockup uses screenshots w/ graphics added where things changed, so if you're confused why some text looks odd—well, that's why!)
- [ ] Separate single screen into 2 distinct states: "Preparing" vs "Ready"
- For _Preparing_
     - [ ] `H1` text to say 'Preparing to [print or export]'
     - [ ] Spinner at top-left instead of device icon
     - [ ] 'Continue' button is disabled
- For _Ready_
     - [ ] `H1` text to say `Ready to [print or export]`
     - [ ] Show device icon in place of spinner at top-left of screen
     - [ ] Enabled "Continue" button

## Etc
This issue is only for the functionality and two words of text, cited above. Any other inconsistencies between text between Zeplin and what's in the code, are not included in this issue.

### LINKED ISSUE #928: Encrypted drive auth screen: active state
_Child issue w/in #910_

## This issue's punchlist
_[Zeplin, here](https://zpl.io/bWEKZDj)_
(note: this mockup uses screenshots w/ graphics added where things changed, so if you're confused why some text looks odd—well, that's why!)
- [ ] Implement active state spinner w/ button styling
- [ ] Disable Password field when actively submitting pane for auths
  - This includes "fading" the label for the text field, and the bullets or text in the field
- [ ] When returning an error'd screen, return the screen to its state w/o the error

### PR DESCRIPTION
**Pull request by ntoll**
# Description

This is a work in progress (WiP).

**NEEDS UX FEEDBACK / CHECKING**

Fixes #927 Fixes #928 Fixes #912.

OK... here's a screenie of the work so far.

active_state

Things to note:

* The spinner @ninavizz has supplied doesn't do infinite looping. I've used the GIMP to re-export the GIF *with* infinite looping and it looks like this. I just want to check you're OK with the result.

activate_state2

* What you're seeing is my "pretend" solution where I put the "CONTINUE" button into the animated state upon dialog opening. My solution is generic enough that all such dialogs (print and usb export) will be able to start and stop this animation at the appropriate points in the dialog's lifecycle. These screenies are just for illustrative purposes.

* These dialogs behave very weird for me -- I'm unable to ALT-TAB away to another window (such as the one I'm using to record my screenies). This feels to me like a UX bad smell.

Please can @ninavizz / @eloquence feedback on the **look** of the work done so far and, once OK I'll make the necessary changes so rather than animating on dialog start, it comes into effect at the appropriate point in the lifecycle of the dialog.

# Test Plan

I've added appropriate unit tests for the new methods in the base `FramelessDialog` class. Please can @ninavizz @eloquence provide UX/UI feedback on this WiP.

**I've not tested on Qubes.**

# Checklist

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:

 - [ ] I have tested these changes in the appropriate Qubes environment
 - [X] I do not have an appropriate Qubes OS workstation set up (the reviewer will need to test these changes)
 - [ ] These changes should not need testing in Qubes

If these changes add or remove files other than client code, packaging logic (e.g., the AppArmor profile) may need to be updated. Please check as applicable:

 - [ ] I have submitted a separate PR to the [packaging repo](https://github.com/freedomofpress/securedrop-debian-packaging)
 - [X] No update to the packaging logic (e.g., AppArmor profile) is required for these changes
 - [ ] I don't know and would appreciate guidance

### CODE REVIEW

**Comment by ninavizz** [comment]
Hi Nicholas! Quick/initial feedback points:
- There's no outline on the `Active State` button. The fill should also be `#f1f1f6` per the Zeplin.
- The spinner size is also shown in Zeplin. If there's a dev reason that cannot be attained, we should troubleshoot that. It's currently too small to be clearly visible... and it also visually "stutters."

My own software on this end is set to generate infinitely looping animations, so @creviera since we're on the same timezone... maybe you or @eloquence could look into this with me? I know they aren't spinning in the GitHub previews, but I did not expect they would also fail to spin in the code. :/

**Comment by creviera** [comment]
> These dialogs behave very weird for me -- I'm unable to ALT-TAB away to another window (such as the one I'm using to record my screenies). This feels to me like a UX bad smell.

The dialogs should not set the `WindowFlag.Popup` as we discussed in the last UX meeting, but that change will need to be tested in Qubes and remember that we will see the window frame. If this is getting in your way of working on active state for refresh, please let me know and we can prioritize the dialog change from qt popup to regular old window dialog with its own close button.

File: securedrop_client/gui/widgets.py [hunk 1]
```diff
@@ -2510,6 +2515,21 @@ def center_dialog(self):
         y_center = (application_window_size.height() - dialog_size.height()) / 2
         self.move(x + x_center, y + y_center)
 
+    def animate_activestate(self):
+        self.continue_button.setIcon(QIcon(self.animation.currentPixmap()))
```
Comment (creviera) [inline securedrop_client/gui/widgets.py]: Just a note: If we use `setIcon` to add an icon to the button, it will appear smaller than the design, but I think this is a good first step. Eventually we'll need to make our own QPushButton class that allows for larger icons, also see: https://github.com/freedomofpress/securedrop-client/pull/679#issuecomment-570636850

**Comment by ninavizz** [comment]
Additional thing I also just noticed @ntoll: the screen shown in this PR never gets a spinner on the button. Only in the top left corner. Just as the mockups in Zeplin show. This is the one screen that shows the primary button in a regular disabled state, whereas the rest of the screens in the print and export flows, should show a spinner active-state button. :)

File: securedrop_client/gui/widgets.py (hunk 1 above)
Comment (ntoll) [reply securedrop_client/gui/widgets.py]: Preach it sister..! :slightly_smiling_face: 

Small steps, etc...

**Comment by ntoll** [comment]
> This is the one screen that shows the primary button in a regular disabled state, whereas the rest of the screens in the print and export flows, should show a spinner active-state button. :)

As I said, what you're seeing is my "pretend" implementation screen so you can see the spinner in-situ but on the wrong screen (Qubes testing is hard for me, as also mentioned).

I'll fix the UI nitpicks mentioned above and ensure the spinner only happens at the right moments in the workflow.

**ATTENTION** someone (not me?) will need to test this in Qubes.

**Comment by ntoll** [comment]
>  * There's no outline on the `Active State` button. The fill should also be `#f1f1f6`
>  per the Zeplin.

The background IS `#f1f1f6` per the Zeplin and I'll remove the outline. Nice spot.

>   * The spinner size is also shown in Zeplin. If there's a dev reason that cannot be attained, we
> should troubleshoot that. It's currently too small to be clearly visible... and it also visually "stutters."

Stuttering is due to my GIF capture. It's smooth on my desktop.

I should be able to update the spinner size on the button in my code.

> * This spinner looks much too thin, too. Is it the correct one that was spec'd for just the button, 
> and not for the top-left corner of this screen?

OK... there have been so many spinner images I'll need to check. In any case, I'll endeavour to use the "wider" version.

> My own software on this end is set to generate infinitely looping animations, so @creviera since
> we're on the same timezone... maybe you or @eloquence could look into this with me? I know they
> aren't spinning in the GitHub previews, but I did not expect they would also fail to spin in the code.
> :/

None of the spinners infinitely loop for me. To fix this I import them into GIMP and re-export them with the "infinitely loop" flag set to true. Then they work.

**Comment by ntoll** [comment]
Visual updates to the active-state button:

- [x] Confirm correct background.
- [x] Correct `QSize(21, 21)` for the `QIcon` associated with the button (using the dimensions from the Zeplin).
- [x] Remove border from active button.

wide_activestate

Need to address the spinner in the top left corner.

**Comment by ninavizz** [comment]
Hey Nicholas! Sorry, I didn't realize Qubes was required to do this on a different screen... :D

Could the spinner be increased in size by ~120%? It still appears too small, even though I don't doubt you checked it correctly against the Zeplin.

**Comment by ntoll** [comment]
@ninavizz you tell me the pixel dimensions you want, and I'd be happy to update. :smile:

**Comment by ntoll** [comment]
May I suggest a nice round number like 32x32..?

**Comment by ninavizz** [comment]
As it's a "proportional" issue... how tall are the buttons in the code?

**Comment by ninavizz** [comment]
Punting on refining the size of the spinner in this PR, per #945. This looks great and gets my signoff, @ntoll! :)

**Comment by ntoll** [comment]
OK folks, this is getting close to being over the line. Here's what I've done:

* Correct spinners.
* Methods on the base class which starts and stops two sorts of animation: activestate (for the primary activity button) and header (for the large icon in the header).
* Various minor CSS changes to make ^^^ look pertty.

I've also added the appropriate calls to the export file pane so that the activestate starts and stops at the appropriate times in the workflow.

**PROBLEM** I can't make sense of [this Zeplin](https://app.zeplin.io/project/5c807ea562f734bd2756b243/screen/5e6b0fd182a344183b46b86a) to know when to start/stop the header spinner. The dialogs shown don't match with what's currently on master. For instance, there's no "Ready to Export" stage (as far as I can see -- grepping for the text doesn't throw up anything).
Some guidance here would be most welcome. To be explicit it'd be wonderful if I could see dialogs that actually look like the current state of master (so I can see where I'm supposed to change) and a full account of the state changes for the print related dialog.

Basically, the heavy lifting is done, but I'm not sure where to activate / deactivate the various animations at the various points in the state changes (and this isn't clear from the Zeplins which don't accurately follow what's currently in master). Once I have this information I'd love to finish this all off.

Finally, once done, this'll need testing in Qubes.

**Comment by ninavizz** [comment]
Hey friend! Apologies my Zeplin was unclear. :/

Stuff was split between the Issue #927 and the Zeplin... but the most significant bit was in the Issue itself, which is that probably a separate PR will need to be created to create that second state. You are correct, that second state does not currently exist. @creviera or @eloquence advisement on how to split this between Nicholas and Qubes-capable peeps?

**Comment by ntoll** [comment]
OK... so since I'm not developing on Qubes I'll need some guidance for when to:

* activate the spinner in the export dialog (I could guess by name, but would like to know for sure). See [this Zeplin](https://app.zeplin.io/project/5c807ea562f734bd2756b243/screen/5e6b0fd182a344183b46b86a) for what I mean. @creviera your name was mentioned as someone who could point this out.
* I should also update the label one the export is ready (so it matches the final state in the Zeplin inked to above).

Then we'll be ready for review. Hurrah..!

**Comment by creviera** [comment]
Just poking my head in here and looks like active state icon on the "Continue" button doesn't go away once the button is enabled. You can test this on non-qubes by applying this diff (check it out on master and your branch to see how before the button would enabled and show the "Continue" text):

```diff
--- a/securedrop_client/logic.py
+++ b/securedrop_client/logic.py
@@ -668,7 +668,9 @@ class Controller(QObject):
         logger.info('Running printer preflight check')
 
         if not self.qubes:
-            self.export.printer_preflight_success.emit()
+            import threading
+            threading.Timer(3, self.export.printer_preflight_success.emit).start()
+            # self.export.printer_preflight_success.emit()
             return
 
         self.export.begin_printer_preflight.emit()
@@ -680,7 +682,9 @@ class Controller(QObject):
         logger.info('Running export preflight check')
 
         if not self.qubes:
-            self.export.preflight_check_call_success.emit()
+            import threading
+            threading.Timer(3, self.export.preflight_check_call_success.emit).start()
+            # self.export.preflight_check_call_success.emit()
             return
 
         self.export.begin_preflight_check.emit()
```

**Comment by ninavizz** [comment]
@ntoll Random clarification...
1. All of the things in this PR, are for _both_ the `Export` and `Print` flows, which begin on near-identical screens. Written content is different, the "Ready" state device icon is different, but that's mostly it. 
2. Per our Slack DM, that first screen never gets the `active spinner` on its Continue/primary button, but the rest of the primary buttons in both flows do. That first screen gets a disabled "Continue" primary button. 
3. I'll be around, later—SMS if u need, will be binge-watching Better Call Saul late tonite!

**Comment by creviera** [comment]
In case it's helpful, here's how you can test each step for both print and export for both failure and success states without QubesOS:

```diff
--- a/securedrop_client/logic.py
+++ b/securedrop_client/logic.py
@@ -43,7 +43,7 @@ from securedrop_client.api_jobs.uploads import SendReplyJob, SendReplyJobError,
     SendReplyJobTimeoutError
 from securedrop_client.api_jobs.updatestar import UpdateStarJob, UpdateStarJobException
 from securedrop_client.crypto import GpgHelper
-from securedrop_client.export import Export
+from securedrop_client.export import Export, ExportError
 from securedrop_client.queue import ApiJobQueue
 from securedrop_client.sync import ApiSync
 from securedrop_client.utils import check_dir_permissions
@@ -668,7 +668,10 @@ class Controller(QObject):
         logger.info('Running printer preflight check')
 
         if not self.qubes:
-            self.export.printer_preflight_success.emit()
+            import threading
+            threading.Timer(3, lambda: self.export.printer_preflight_failure.emit(ExportError('errrrror'))).start()
+            # threading.Timer(3, self.export.printer_preflight_success.emit).start()
+            # self.export.printer_preflight_success.emit()
             return
 
         self.export.begin_printer_preflight.emit()
@@ -680,7 +683,10 @@ class Controller(QObject):
         logger.info('Running export preflight check')
 
         if not self.qubes:
-            self.export.preflight_check_call_success.emit()
+            import threading
+            threading.Timer(3, self.export.preflight_check_call_success.emit).start()
+            # threading.Timer(3, lambda: self.export.preflight_check_call_failure.emit(ExportError('errrrror'))).start()
+            # self.export.preflight_check_call_success.emit()
             return
 
         self.export.begin_preflight_check.emit()
@@ -699,7 +705,10 @@ class Controller(QObject):
             return
 
         if not self.qubes:
-            self.export.export_usb_call_success.emit()
+            import threading
+            threading.Timer(3, lambda: self.export.export_usb_call_failure.emit(ExportError('errrrror'))).start()
+            # threading.Timer(3, self.export.export_usb_call_success.emit).start()
+            # self.export.export_usb_call_success.emit()
             return
 
         self.export.begin_usb_export.emit([file_location], passphrase)
```

**Comment by ntoll** [comment]
OK... this was more fiddly than I imagined since the buttons were moving around the UI because the `QSize` was not what it ought to be and I simply had to use my Eyeball Mk.1 to judge what the actual dimensions were and figure out the CSS. These truly are artisnal, hand crafted, organic spinner buttons. :wink: Look at how hipster am I..? (I still hate CSS with a vengeance though... I must have spent years faffing about with it so it's "just so" for no logical reason.) 

@creviera your input was invaluable in allowing me to see how it'd work on Qubes. Thank you. Screenies below...

Export success:

export_to_usb_spinners

Export fail (the error message is me testing... please ignore the content):

export_to_usb_fail

Print success:

print_spinner

Print fail (behaviour is as currently is, error message is me testing):

print_spinner_error


The most important aspect of these screenies is that the spinners and other active state doodahs appear in the right place and stop at the appropriate point.

This last part was about as fiddly than the string section of the Boston Pops Orchestra playing selections from "Fiddler on the Roof". :violin: :wink:

**Comment by ntoll** [comment]
OK... putting aside the build fail (due to CI infrastructure problem AFAICT), this is ready for review. Please see screenies in previous comment. Many thanks to @ninavizz and @creviera for all the help getting this over the line.

**Comment by creviera** [comment]
I think this looks great @ntoll! I'll try to see what's going on with the functional test failure for starring... looks like we're getting `sqlite3.OperationalError: unable to open database file`

**Comment by redshiftzero** [comment]
that test error was a flake (cross-linking #956), restarted and CI now passes (`test-buster` job will appear as succeeded in ~10 mins)

**Comment by eloquence** [comment]
This is marked to resolve #928. Does it in fact fully do so, including these two items?

- [ ] Disable Password field when actively submitting pane for auths
- [ ] When returning an error'd screen, return the screen to its state w/o the error

**Comment by ntoll** [comment]
OK... here's a couple of screenies to show how this now looks in light of @eloquence's pointing out I'd missed a couple of points; disabling the password field on submit and ensuring the error message is displayed as expected and then dimmed on re-submit:

error_dim

**Comment by ninavizz** [comment]
Ohhh that is lovely! <3

**Comment by ntoll** [comment]
Hurrah. Thank you for all your help folks...! :sun_with_face:
